src/main/haskell/language-kore/src/Data/Kore/ASTPrettyPrint.hs

```diff
@@ -0,0 +1,471 @@
+2. Makes sure that the code still works when the field order changes
+   (done by using the something<n> accessors).
+3. Makes sure that we notice when field names change (again, done by using
+   the something<n> accessors).
+
+However, this generates the HLint error disabled above.
+-}
+
+class FromString a where
+    fromString :: String -> a
+
+class (FromString w, MonadWriter w m, MonadReader Int m)
+    => PrettyPrintOutput w m where
+    write :: String -> m ()
+    write s = tell (fromString s)
+
+    betweenLines :: m ()
+    betweenLines = do
+        indent <- reader (`replicate` ' ')
+        write "\n"
+        write indent
+
+    withIndent :: Int -> m() -> m ()
+    withIndent n = local (+n)
+
+class PrettyPrint a where
+    prettyPrint :: PrettyPrintOutput w m => a -> m ()
+
+{-  Unparse to string instance
+-}
+type StringPrettyPrinter = WriterT ShowS (Reader Int)
+
+instance FromString ShowS where
+    fromString = showString
+
+instance PrettyPrintOutput ShowS StringPrettyPrinter where
+
+-- TODO: Do I need this?
+stringPrettyPrint :: PrettyPrint a => a -> StringPrettyPrinter ()
+stringPrettyPrint = prettyPrint
+
+prettyPrintToString :: PrettyPrint a => a -> String
+prettyPrintToString a = showChain ""
+  where
+    writerAction = stringPrettyPrint a
+    readerAction = execWriterT writerAction
+    showChain = runReader readerAction 0
+
+{- utility functions -}
+
+writeOneFieldStruct
+    :: (PrettyPrintOutput w m, PrettyPrint a)
+    => String -> a -> m()
+writeOneFieldStruct name content = do
+    write "("
+    write name
+    write " "
+    prettyPrint content
+    write ")"
+
+writeOneFieldStructNewLine
+    :: (PrettyPrintOutput w m, PrettyPrint a)
+    => String -> a -> m()
+writeOneFieldStructNewLine name content = do
+    write "("
+    write name
+    withIndent 4 (prettyPrint content)
+    write ")"
+
+writeFieldOneLine
+    :: (PrettyPrintOutput w m, PrettyPrint a)
+    => String -> (b->a) -> b -> m()
+writeFieldOneLine fieldName field object = do
+    write fieldName
+    write " = "
+    prettyPrint (field object)
+
+writeListField
+    :: (PrettyPrintOutput w m, PrettyPrint a)
+    => String -> (b->a) -> b -> m()
+writeListField fieldName field object = do
+    write fieldName
+    write " ="
+    prettyPrint (field object)
+
+writeFieldNewLine
+    :: (PrettyPrintOutput w m, PrettyPrint a) => String -> (b->a) -> b -> m()
+writeFieldNewLine fieldName field object = do
+    write fieldName
+    write " ="
+    withIndent 4
+        (betweenLines >> prettyPrint (field object))
+
+writeAttributesField
+    :: (PrettyPrintOutput w m)
+    => String
+    -> Attributes
+    -> m()
+writeAttributesField fieldName (Attributes []) = do
+    write fieldName
+    write " = Attributes []"
+writeAttributesField fieldName attributes = do
+    write fieldName
+    write " ="
+    withIndent 4
+        (betweenLines >> prettyPrint attributes)
+
+writeStructure :: PrettyPrintOutput w m => String -> [m()] -> m()
```

Comment:
also, you should be consistent with `m()` vs. `m ()`.  Probably the latter is better.